server/middlewares.go

```diff
@@ -45,3 +46,19 @@ func serverTimeoutMW(initCtx context.Context) endpoint.Middleware {
 		}
 	}
 }
+
+// newCtxInjectMW must be placed at the end
+func newCtxInjectMW() endpoint.Middleware {
```

Comment:
The middleware mechanism is mainly used by external users, and I strongly oppose writing internal code in fragmented snippets that require additional function calls for every single request.

Reply:
Yes, this is a workaround. When the streaming API with ctx is published, this part would be removed.
If we do not use middleware mechanism, we could:
- modify the logic in invokeHandleEndpoint

```
methodInfo := svcInfo.MethodInfo(methodName)
if methodInfo.IsStreaming() {
    if sArgs, ok := args.(*streaming.Args); ok {
        sArgs.Stream = &contextStream{
	    Stream: sArgs.Stream,
	    ctx:    ctx,
	}
    }
}
implHandlerFunc := methodInfo.Handler()
```
But this judgement would affect other ping-pong thrift requests.
- expose SetContext for Stream interface and put it in internal pkg

When new Stream interface is published, this method seems a bit redundant.